### PR TITLE
Update custom-functions guide

### LINKED ISSUE #968: Custom functions guide is outdated
### Description

From #965 we learned that the guide gives inaccurate information. It must be updated.

Some changes in the production code might also be needed. E.g. a custom function should work without `parameters` provided in the metadata:
https://github.com/handsontable/hyperformula/blob/38e31a702b8b6ffa237a750a6c1c0164f67d432e/src/interpreter/plugin/FunctionPlugin.ts#L48

### Steps to reproduce

1. Try to implement a function following the guide
2. Fail

### Demo

Included in #965

### Links
#965
#779
#957

### PR DESCRIPTION
**Pull request by sequba**
### Context

Changes to the `custom-functions` guide:
- start the guide with a step-by step tutorial about implementing the simplest possible function
- make it clear that defining translations is a required step
- encourage using `runFunction` helper method
- convert the code examples from Typescript to Javascript
- fix minor bugs in code examples

Other changes:
- fixa bug in production code: `implementedFunctions.parameters` should be optional
- add a few unit tests not related to custom functions. I wrote these tests while investigating other issues.

TODO:
- [ ] update hyperformula-demos

### How has this been tested?

- [ ] verify code examples in CodeSandbox
- [ ] run documentation locally and make sure it renders correctly 

### Types of changes

- [x] Docs

### Related issue(s):
1. Fixes #968 
2. https://github.com/handsontable/hyperformula/issues/965
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

File: src/interpreter/plugin/FunctionPlugin.ts [hunk 1]
```diff
@@ -355,7 +355,7 @@ export abstract class FunctionPlugin implements FunctionPluginTypecheck<Function
     metadata: FunctionMetadata,
     fn: (...arg: any) => InterpreterValue,
   ) => {
-    let argumentDefinitions: FunctionArgument[] = metadata.parameters!
+    let argumentDefinitions: FunctionArgument[] = metadata.parameters || []
```
Comment (sequba) [inline src/interpreter/plugin/FunctionPlugin.ts:358]: This is a bugfix. Our documentation says that `parameters` field is optional, but this function fails when it's not defined.

**Review by kirszenbaum — CHANGES_REQUESTED** [review]
@sequba – I added my suggestions in #1010

**Comment by sequba** [comment]
@kirszenbaum suggestions applied.

File: docs/guide/custom-functions.md [hunk 2]
```diff
@@ -1,12 +1,12 @@
 # Custom functions
 
-Expand the function library of your application, by adding custom functions.
+Expand the function library of your application by adding custom functions.
```
Comment (kirszenbaum) [inline docs/guide/custom-functions.md:3]:
```suggestion
Expand the function library of your application, by adding custom functions.
```

Let's leave the comma here.

File: docs/guide/custom-functions.md [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
 [[toc]]
 
 ## Add a custom function
 
-As an example, let's create a function that returns the number of letters in the word "HyperFormula".
+For example, let's create a function that returns the number of letters in the word "HyperFormula".
```
Comment (kirszenbaum) [inline docs/guide/custom-functions.md:9]:
```suggestion
As an example, let's create a function that returns the number of letters in the word "HyperFormula".
```

File: docs/guide/custom-functions.md [hunk 4]
```diff
@@ -87,7 +87,7 @@ CountHF.translations = {
 
 ### 4. Implement your function's logic
 
-In your function plugin, add a method that implements your function's calculations. Your method needs to:
+Add a method that implements your function's calculations in your function plugin. Your method needs to:
```
Comment (kirszenbaum) [inline docs/guide/custom-functions.md:90]:
```suggestion
In your function plugin, add a method that implements your function's calculations. Your method needs to:
```

Location comes first, action second.

File: docs/guide/custom-functions.md [hunk 5]
```diff
@@ -209,7 +209,7 @@ You can set the following options for your function:
 
 You can set rules for your function's argument validation.
 
-In your function plugin, in the static `implementedFunctions` property, next to other options add an array called `parameters`:
+In your function plugin, in the static `implementedFunctions` property, next to other options, add an array called `parameters`:
```
Comment (kirszenbaum) [inline docs/guide/custom-functions.md:212]:
```suggestion
In your function plugin, in the static `implementedFunctions` property, next to other options add an array called `parameters`:
```

File: docs/guide/custom-functions.md [hunk 6]
```diff
@@ -243,9 +243,9 @@ You can set the following argument validation options:
 
 #### Handling missing arguments
 
-Both the `defaultValue` and `optionalArg` options let you decide what happens when a user doesn't pass enough valid arguments to your custom function.
+The `defaultValue` and `optionalArg` options let you decide what happens when a user doesn't pass enough valid arguments to your custom function.
```
Comment (kirszenbaum) [inline docs/guide/custom-functions.md:246]:
```suggestion
Both the `defaultValue` and `optionalArg` options let you decide what happens when a user doesn't pass enough valid arguments to your custom function.
```

**Review by kirszenbaum — CHANGES_REQUESTED** [review]

**Review by kirszenbaum — APPROVED** [review]

**Pull request merged**